docs: Add quick start for the Docker provider

dlipovetsky opened the pull request:
**What this PR does / why we need it**:
Add Docker provider to the book quick start.

/cc @chuckha

chuckha commented on docs/book/src/user/quick-start.md:
```diff
@@ -219,6 +291,49 @@ spec:
+  initConfiguration:
+    nodeRegistration:
+      kubeletExtraArgs:
+        eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
```
I don't usually set this, but i haven't been doing much dev work on capd lately -- is this a kind thing?

dlipovetsky replied:
Caught red-handed! I copied this from https://github.com/kubernetes-sigs/cluster-api-provider-docker/blob/master/examples/simple-cluster.yaml, which was updated in https://github.com/kubernetes-sigs/cluster-api-provider-docker/pull/248. :smile: 

That PR doesn't explain these changes, but here's my take:

The lower the thresholds for hard eviction, the fewer total resources are required, and more Pods can be scheduled. WIth these thresholds, Pods can completely starve the node of resources. The default thresholds are higher to prevent this from happening.

Since Docker provider is running on one machine, and isn't intended for production, it might be reasonable to allow scheduling more Pods at the risk of starving the node (a container) of resources.

I'm happy to add an explanation to the doc or manifest.
 
@erwinvaneyk WDYT?

erwinvaneyk replied:
To complete the circle, I saw this in @chuckha's https://github.com/chuckha/capi-dev setup. 

Since I ran into eviction issues on my own machine (which had only 20-25% remaining disk space) due to the issue that @dlipovetsky described, I think it is reasonable to add here.

chuckha replied:
perfect. sgtm!

dlipovetsky replied:
Thanks @erwinvaneyk. I'll add an explanatory comment.

chuckha commented on docs/book/src/user/quick-start.md:
```diff
@@ -219,6 +291,49 @@ spec:
+  clusterConfiguration:
+    controllerManager:
+      extraArgs:
+        enable-hostpath-provisioner: "true"
```
same with this, i'm guessing this is a requirement for kind, but i'm not sure

dlipovetsky replied:
Adding explanatory comment

chuckha commented on docs/book/src/user/quick-start.md:
```diff
@@ -339,6 +454,24 @@ kubectl --kubeconfig=./capi-quickstart.kubeconfig \
+<aside class="note warning">
+<h1>Action Required (Docker provider v0.2.0 only)</h1>
+
+A [known issue](https://github.com/kubernetes-sigs/kind/issues/891) affects Calico with the Docker provider v0.2.0. After you deploy Calico, apply this patch to work around the issue:
```
perhaps we can use a different CNI? maybe weave?

detiber replied:
Weave would allow for removing the need to override Cluster.Spec.ClusterNetwork as well.

dlipovetsky replied:
Filed https://github.com/kubernetes-sigs/cluster-api/issues/1533

detiber commented on docs/book/src/user/quick-start.md:
```diff
@@ -7,10 +7,51 @@ In this tutorial we'll cover the basics of how to use Cluster API to create one
+  - role: control-plane
+    extraMounts:
+      - hostPath: /var/lib/docker
+        containerPath: /var/lib/docker
```
Is `/var/lib/docker` actually needed here? My testing to date has just mounted `/var/run/docker.sock` only.

dlipovetsky replied:
Thanks. Removing.

detiber commented on docs/book/src/user/quick-start.md:
````diff
@@ -7,10 +7,51 @@ In this tutorial we'll cover the basics of how to use Cluster API to create one
+  ```bash
+  kind create cluster --name=clusterapi
+  export KUBECONFIG="$(kind get kubeconfig-path --name="clusterapi")"
+  ```
````
Not a blocker, but it would be nice if we could reference this block form other tabs rather than replicating in multiple places.

vincepri replied:
I'd prefer this as well, maybe we can have a special KIND configuration section when you have to pick your infrastructure provider?

detiber commented on docs/book/src/user/quick-start.md:
```diff
@@ -219,6 +291,49 @@ spec:
+  labels:
+    cluster.x-k8s.io/control-plane: "true"
+    cluster.x-k8s.io/cluster-name: "capi-quickstart"
+    set: controlplane
```
I'd probably remove the `set` label here, since it isn't necessary

dlipovetsky replied:
Removing, thanks

detiber commented on docs/book/src/user/quick-start.md:
```diff
@@ -417,6 +550,69 @@ spec:
+  name: capi-quickstart-worker
+spec:
+  template:
+    spec: {}
```
Yuck, I'm wondering if we should make some of these fields optional rather than required.

vincepri commented on docs/book/src/user/quick-start.md:
```diff
@@ -39,7 +80,7 @@ kubectl create -f {{#releaselink gomodule:"sigs.k8s.io/cluster-api-bootstrap-pro
 #### Install Infrastructure Provider Components
 
-{{#tabs name:"tab-installation-infrastructure" tabs:"AWS,vSphere"}}
+{{#tabs name:"tab-installation-infrastructure" tabs:"AWS,Docker,vSphere"}}
```
Let's keep Docker last after all cloud providers, wdyt?

dlipovetsky replied:
I chose alphabetical order, because I assume people are used to finding items that way.

Do you want to place Docker last because cloud-backed providers are meant for production, and Docker?

vincepri replied:
Yeah, if that makes sense, we can add a comment to keep non cloud providers at the end

dlipovetsky replied:
I would prefer to keep alphabetical order, because I think newcomers will be able to distinguish the providers by name alone. But I am biased, because I know the various providers. It would be great to have input from newcomers :smile:

randomvariable replied:
It's only three of them at the moment.  Let's revisit once we get crowded.

ncdc commented on docs/book/src/user/quick-start.md:
```diff
@@ -7,10 +7,51 @@ In this tutorial we'll cover the basics of how to use Cluster API to create one
+
+</aside>
+
+  Because the Docker provider needs to access Docker on the host, a custom kind cluster configuration is required:
```
Do we need a note somewhere that says it doesn't work with Docker for Mac? @chuckha

dlipovetsky replied:
@erwinvaneyk Your thoughts?

erwinvaneyk replied:
Both v1alpha1 and v1alpha2 work on docker-for-mac. In both cases you need to adapt the kubeconfig a bit (with v1alpha2 it is a bit more tricky than with v1alpha1). I don't recall the exact steps though, will need to look them up later.

chuckha replied:
Yeah, we need to fix this issue. The steps for os x are to look at the exposed port that maps to the api server on the load balancer and update your kubeconfig. I can provide some instructions today to make this complete

chuckha replied:
note: this is fine to fix after merge

dlipovetsky replied:
I'll leave a comment that MacOS instructions are forthcoming

dlipovetsky replied:
Filed https://github.com/kubernetes-sigs/cluster-api/issues/1534

randomvariable commented:
How many of the outstanding comments are blockers? Can we create issues and fast follow on improvements (e.g. swapping Calico for Weave)?

detiber commented:
@randomvariable I'm good with issues and followups for the comments I raised.

vincepri commented:
We should target the `release-0.2` branch for this PR and forward port the content once we have CAPD automation in place now that it has been merged on master

dlipovetsky commented:
Thanks everyone for your feedback! I got busy and wasn't able to look at the PR. Will look over comments now and will address minor things, and happy to file issues for major things, or where we don't have consensus right now.

dlipovetsky commented:
Squashed and rebased on changes from #1318.

I removed the redundant "Prerequisites" section the top of the quick start, because the very same section appears at the top of the Installation doc, which is included in the quick start doc.

dlipovetsky commented:
Added the 📖 unicode symbol to the commit subject :-)

dlipovetsky commented:
Removed a line that got left during the rebase.

dlipovetsky commented:
@detiber Won't a CAPD quick start also make sense in v1a3 (on master)?

detiber commented:
> @detiber Won't a CAPD quick start also make sense in v1a3 (on master)?

Yes, but without a release yet, I'm not sure what changes would be needed to make it work for master/v1alpha3.

dlipovetsky commented:
Rebased on `release-0.2`

dlipovetsky commented:
Argh, I pushed prior to switching the base... Now I need to clean up these extra commits, I guess

dlipovetsky commented:
Nevermind, I had commits from master in my branch. Removed now.

wfernandes commented:
@dlipovetsky I was wondering what had happened. Thank you for this PR. I'm using it right now to try and quick start CAPD on my local mac machine.

dlipovetsky commented:
/hold Don't merge, again the "install infra provider" section for Docker is missing :(

dlipovetsky commented:
/hold cancel Fixed.

randomvariable commented:
/lgtm

dlipovetsky commented:
/hold cancel

(Didn't I cancel it already?)
